File: backend/app/schemas/note.py
from datetime import datetime

from pydantic import BaseModel


class NoteBase(BaseModel):
    title: str
    content: str | None = None
    sidenote: str | None = None
    parent_id: str | None = None


class NoteCreate(NoteBase):
    pass


class NoteUpdate(BaseModel):
    title: str | None = None
    content: str | None = None
    sidenote: str | None = None
    parent_id: str | None = None
    position: int | None = None


class NoteReorder(BaseModel):
    """Payload for reordering/moving a note"""
    parent_id: str | None = None  # New parent (null for root)
    position: int  # New position within parent


class NoteResponse(NoteBase):
    id: str
    position: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/types/note.ts
export interface Note {
  id: string;
  title: string;
  content: string | null;
  sidenote: string | null;
  parent_id: string | null;
  position: number;
  created_at: string;
  updated_at: string;
}

export interface NoteCreate {
  title: string;
  content?: string | null;
  sidenote?: string | null;
  parent_id?: string | null;
}

export interface NoteUpdate {
  title?: string;
  content?: string | null;
  sidenote?: string | null;
  parent_id?: string | null;
  position?: number;
}

export interface NoteReorder {
  parent_id: string | null;
  position: number;
}

// Tree structure for sidebar
export interface NoteTreeNode extends Note {
  children: NoteTreeNode[];
}
